File: Java/src/_0075_Sort_Colors/Solution.java
package _0075_Sort_Colors;

import _0000_study._common.Output;

/**
 * 时间复杂度：O(n)
 * 空间复杂度：O(1)
 * 计数排序
 */
class Solution {

    public void sortColors(int[] nums) {

        int[] cnt = new int[3];  // 统计0,1,2的次数

        for (int num : nums) cnt[num]++;

        int index = 0;
        for (int i = 0; i < cnt.length; i++)
            for (int j = 0; j < cnt[i]; j++)
                nums[index++] = i;

    }

    public static void main(String[] args) {

        int[] nums = {2, 0, 2, 1, 1, 0};
        (new Solution()).sortColors(nums);
        Output.OutputBasicArray1D(nums);
    }
}

File: Python/_0075_Sort_Colors/Solution.py
# coding=utf-8
# Date: 2021/5/26 16:58
from typing import List


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = 0; j = 0; k = len(nums) - 1
        while i <= k:
            if nums[i] == 0:
                nums[j], nums[i] = nums[i], nums[j]
                i += 1; j += 1
            elif nums[i] == 2:
                nums[i], nums[k] = nums[k], nums[i]
                k -= 1
            else:
                i += 1


if __name__ == "__main__":
    pass
